File: output/opt3001-driver/src/python/opt3001.py
import smbus2
import time

class OPT3001:
    # I2C address
    I2C_ADDR = 0x44
    
    # Register addresses
    REG_RESULT = 0x00
    REG_CONFIG = 0x01
    REG_LOW_LIMIT = 0x02
    REG_HIGH_LIMIT = 0x03
    REG_MANUFACTURER_ID = 0x7E
    REG_DEVICE_ID = 0x7F
    
    # Configuration settings
    MODE_SHUTDOWN = 0x0000
    MODE_SINGLE_SHOT = 0x2000
    MODE_CONTINUOUS = 0x4000
    
    INT_TIME_100MS = 0x0000
    INT_TIME_800MS = 0x0800
    
    def __init__(self, bus_number=1):
        self.bus = smbus2.SMBus(bus_number)
        self.configure(self.MODE_CONTINUOUS | self.INT_TIME_100MS)
    
    def read_lux(self):
        """Read the light level in lux"""
        # Read result register
        data = self.bus.read_i2c_block_data(self.I2C_ADDR, self.REG_RESULT, 2)
        raw_result = (data[0] << 8) | data[1]
        
        # Extract exponent and mantissa
        exponent = (raw_result >> 12) & 0x0F
        mantissa = raw_result & 0x0FFF
        
        # Calculate lux according to datasheet formula
        lux = 0.01 * (1 << exponent) * mantissa
        return lux
    
    def configure(self, configuration):
        """Configure the sensor with the given settings"""
        # Split 16-bit configuration into two bytes
        msb = (configuration >> 8) & 0xFF
        lsb = configuration & 0xFF
        self.bus.write_i2c_block_data(self.I2C_ADDR, self.REG_CONFIG, [msb, lsb])
    
    def __del__(self):
        self.bus.close()

File: output/opt3001-driver/src/ts/opt3001.ts
import { promisify } from 'util';
import { I2CBus } from 'i2c-bus';

export class Opt3001 {
    private static readonly I2C_ADDR: number = 0x44;
    
    // Register addresses
    private static readonly REG_RESULT: number = 0x00;
    private static readonly REG_CONFIG: number = 0x01;
    
    // Configuration settings
    static readonly MODE_CONTINUOUS: number = 0x4000;
    static readonly INT_TIME_100MS: number = 0x0000;
    
    private bus: I2CBus;
    private readByte: (addr: number, cmd: number) => Promise<number>;
    private writeByte: (addr: number, cmd: number, byte: number) => Promise<void>;
    private readWord: (addr: number, cmd: number) => Promise<number>;
    private writeWord: (addr: number, cmd: number, word: number) => Promise<void>;
    
    constructor(bus: I2CBus) {
        this.bus = bus;
        this.readByte = promisify(bus.readByte.bind(bus));
        this.writeByte = promisify(bus.writeByte.bind(bus));
        this.readWord = promisify(bus.readWord.bind(bus));
        this.writeWord = promisify(bus.writeWord.bind(bus));
        
        // Configure with default settings
        this.configure(Opt3001.MODE_CONTINUOUS | Opt3001.INT_TIME_100MS);
    }
    
    async readLux(): Promise<number> {
        // Read result register
        const rawResult = await this.readWord(Opt3001.I2C_ADDR, Opt3001.REG_RESULT);
        
        // Extract exponent and mantissa
        const exponent = (rawResult >> 12) & 0x0F;
        const mantissa = rawResult & 0x0FFF;
        
        // Calculate lux according to datasheet formula
        const lux = 0.01 * (1 << exponent) * mantissa;
        return lux;
    }
    
    async configure(configuration: number): Promise<void> {
        await this.writeWord(Opt3001.I2C_ADDR, Opt3001.REG_CONFIG, configuration);
    }
}
